File: server/encrypted-object/ports.ts
import type { VaultContext } from '../../lib/domain/identity';
import type { VaultPartitionRoute } from '../vault-content/records';
import type {
  DeleteOutboxEntry,
  EncryptedObjectMetadata,
  EncryptedWriteId,
  OpaqueObjectKey,
  PendingEncryptedWrite,
  PrivateObjectDescriptor,
} from './core';
import type { EnvelopeObject } from '../crypto/core';
import type { CryptoObjectRevision } from '../crypto/core';

export type ImmutableObjectPutResult =
  | { readonly kind: 'stored' }
  | { readonly kind: 'already-present' }
  | { readonly kind: 'conflict' };

export type PrivateObjectStoragePort = {
  get(objectKey: OpaqueObjectKey): Promise<Uint8Array | undefined>;
  putIfAbsent(input: {
    readonly objectKey: OpaqueObjectKey;
    readonly bytes: Uint8Array;
    readonly createdAt: number;
  }): Promise<ImmutableObjectPutResult>;
  delete(objectKey: OpaqueObjectKey): Promise<void>;
  list(): Promise<readonly PrivateObjectDescriptor[]>;
};

export type OpaqueObjectKeyGeneratorPort = {
  createObjectKey(): Promise<unknown>;
};

export type IntentReservationResult =
  | { readonly kind: 'reserved'; readonly intent: PendingEncryptedWrite }
  | { readonly kind: 'existing'; readonly intent: PendingEncryptedWrite }
  | { readonly kind: 'conflict' };

export type MetadataCommitResult =
  | { readonly kind: 'applied'; readonly metadata: EncryptedObjectMetadata }
  | { readonly kind: 'not-applied' };

export type EncryptedObjectMetadataRepository = {
  findCurrent(
    object: EnvelopeObject,
  ): Promise<EncryptedObjectMetadata | undefined>;
  findRevision(
    object: EnvelopeObject,
    objectRevision: CryptoObjectRevision,
  ): Promise<EncryptedObjectMetadata | undefined>;
  findByWriteId(
    writeId: EncryptedWriteId,
  ): Promise<EncryptedObjectMetadata | undefined>;
  findIntent(
    writeId: EncryptedWriteId,
  ): Promise<PendingEncryptedWrite | undefined>;
  reserveIntent(
    intent: PendingEncryptedWrite,
  ): Promise<IntentReservationResult>;
  commitIntent(input: {
    readonly intent: PendingEncryptedWrite;
    readonly ciphertextBytes: number;
  }): Promise<MetadataCommitResult>;
  abandonIntent(input: {
    readonly intent: PendingEncryptedWrite;
    readonly requestedAt: number;
  }): Promise<void>;
  listProtectedObjectKeys(): Promise<ReadonlySet<OpaqueObjectKey>>;
  enqueueDelete(input: {
    readonly objectKey: OpaqueObjectKey;
    readonly requestedAt: number;
  }): Promise<void>;
  listReadyDeletes(input: {
    readonly now: number;
    readonly limit: number;
  }): Promise<readonly DeleteOutboxEntry[]>;
  completeDelete(entry: DeleteOutboxEntry): Promise<void>;
  rescheduleDelete(entry: DeleteOutboxEntry): Promise<void>;
};

export type EncryptedObjectRepositoryOpenResult =
  | {
      readonly kind: 'opened';
      readonly route: VaultPartitionRoute;
      readonly repository: EncryptedObjectMetadataRepository;
    }
  | { readonly kind: 'not-found' };

export type EncryptedObjectMetadataDirectory = {
  open(context: VaultContext): Promise<EncryptedObjectRepositoryOpenResult>;
};

File: backend/internal/encryptedobject/ports.go
package encryptedobject

import (
	"context"

	"github.com/fukamu/notes/backend/internal/cryptocontent"
)

type PutResult string

const (
	PutStored         PutResult = "stored"
	PutAlreadyPresent PutResult = "already-present"
	PutConflict       PutResult = "conflict"
)

type DeleteResult string

const (
	DeleteDeleted  DeleteResult = "deleted"
	DeleteNotFound DeleteResult = "not-found"
)

type ObjectStoragePort interface {
	Get(context.Context, ObjectKey) ([]byte, bool, error)
	PutIfAbsent(context.Context, ObjectKey, []byte, int64) (PutResult, error)
	Delete(context.Context, ObjectKey) (DeleteResult, error)
	List(context.Context) ([]PrivateObjectDescriptor, error)
}

type ObjectKeyGeneratorPort interface {
	CreateObjectKey(context.Context) (string, error)
}

type EncryptionPort interface {
	Encrypt(context.Context, cryptocontent.VaultDEKKeyring, cryptocontent.ObjectContext, []byte) (cryptocontent.EnvelopeCiphertext, error)
	Decrypt(context.Context, cryptocontent.VaultDEKKeyring, cryptocontent.ObjectContext, cryptocontent.EnvelopeCiphertext) ([]byte, error)
}

type IntentReservationKind string

const (
	IntentReserved IntentReservationKind = "reserved"
	IntentExisting IntentReservationKind = "existing"
	IntentConflict IntentReservationKind = "conflict"
)

type IntentReservation struct {
	Kind   IntentReservationKind
	Intent PendingWrite
}

type MetadataCommitKind string

const (
	MetadataApplied    MetadataCommitKind = "applied"
	MetadataNotApplied MetadataCommitKind = "not-applied"
)

type MetadataCommit struct {
	Kind     MetadataCommitKind
	Metadata Metadata
}

type MetadataRepository interface {
	FindCurrent(context.Context, ObjectRef) (*Metadata, error)
	FindRevision(context.Context, ObjectRef, cryptocontent.ObjectRevision) (*Metadata, error)
	FindByWriteID(context.Context, WriteID) (*Metadata, error)
	FindIntent(context.Context, WriteID) (*PendingWrite, error)
	ReserveIntent(context.Context, PendingWrite) (IntentReservation, error)
	CommitIntent(context.Context, PendingWrite, int64) (MetadataCommit, error)
	AbandonIntent(context.Context, PendingWrite, int64) error
	ListProtectedObjectKeys(context.Context) (map[ObjectKey]struct{}, error)
	EnqueueDelete(context.Context, ObjectKey, int64) (bool, error)
	ListReadyDeletes(context.Context, int64, int) ([]DeleteOutboxEntry, error)
	CompleteDelete(context.Context, DeleteOutboxEntry) error
	RescheduleDelete(context.Context, DeleteOutboxEntry) error
}
